### PR TITLE
feat: Implement IConnectorV2

### PR DESCRIPTION
**Pull request by AndreasNicolaisen**
Depends on #https://github.com/CluedIn-io/CluedIn.Connector.Common/pull/38

# Description
Work Item ID: [AB#21242](https://dev.azure.com/CluedIn-io/c054b4ae-1dab-43c2-af97-3683c744782f/_workitems/edit/21242)

Implement new connector interface.
This will export incoming and outgoing edges separately, so they can be distinguished from each other.
This will also export properties on edges.
Everything other than the main table is exported via table typed parameters, so the number of edges, edge properties and codes, will not impact the number of parameters used.

# Test approach
New unit tests added, and manually tested.

# Release notes
feat: Implement new connector interface, exporting edges separately, and edge properties.

### CODE REVIEW

File: src/Connector.SqlServer/Utils/TableDefinitions/CodeTableDefinition.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        }
+
+        public static SqlParameter GetPivotKeyParameter(StreamMode mode, SqlConnectorEntityData connectorEntityData)
+        {
+            return new SqlParameter("@EntityId", SqlDbType.UniqueIdentifier) { Value = connectorEntityData.EntityId };
+        }
+
+        public static SqlServerConnectorCommand CreateUpsertCommand(IReadOnlyStreamModel streamModel, SqlConnectorEntityData connectorEntityData, SqlName schema)
+        {
+            var mode = streamModel.Mode ?? StreamMode.Sync; // TODO: Correct fallback?
```
Comment (zvirja) [inline src/Connector.SqlServer/Utils/TableDefinitions/CodeTableDefinition.cs:81]: What is the case when it's empty? 
I think we should make the decision and remove TODO. If not us now, who will know better in future? :)

**Review by zvirja — CHANGES_REQUESTED** [review]
Brilliant work 🤩 Was a pleasure to read your code!!! I love the design of it and how easy it's to follow. Awesome! 🥇 

Now let's hope that harsh reality and performance testing will not force you squash all those descriptive bits into one big query 😟 

I didn't test your code and just did a surface spell-checking review. See my suggestions below.

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 2]
```diff
@@ -22,546 +16,463 @@
...
+                    await _client.ExecuteCommandInTransactionAsync(transaction, command.Text, command.Parameters);
                 }
 
-                await Task.WhenAll(tasks);
                 await transaction.CommitAsync();
             });
         }
 
-        public override async Task EmptyContainer(ExecutionContext executionContext, Guid providerDefinitionId, string id)
+        public override async Task EmptyContainer(ExecutionContext executionContext, IReadOnlyStreamModel streamModel)
```
[the remark this answers is not in the document]
Comment (AndreasNicolaisen) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:270]: Change to execute command for each table.

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 3]
```diff
@@ -22,546 +16,463 @@
...
             });
         }
 
-        public override async Task<IEnumerable<IConnectorContainer>> GetContainers(ExecutionContext executionContext, Guid providerDefinitionId)
+        public override async Task<IEnumerable<IConnectorContainer>> GetContainers(ExecutionContext executionContext, Guid connectorProviderDefinitionId)
         {
-            var config = await base.GetAuthenticationDetails(executionContext, providerDefinitionId);
+            var config = await AuthenticationDetailsHelper.GetAuthenticationDetails(executionContext, connectorProviderDefinitionId);
             await using var connectionAndTransaction = await _client.BeginTransaction(config.Authentication);
             var transaction = connectionAndTransaction.Transaction;
```
[the remark this answers is not in the document]
Comment (AndreasNicolaisen) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:469]: Agreed, changed to use LINQ.

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 4]
```diff
@@ -22,546 +16,463 @@
...
+                sqlCommand.CommandText = commandText;
+                sqlCommand.Transaction = transaction;
+
+                await sqlCommand.ExecuteNonQueryAsync();
                 await transaction.CommitAsync();
             });
         }
 
-        public override async Task<IEnumerable<IConnectorContainer>> GetContainers(ExecutionContext executionContext, Guid providerDefinitionId)
+        public override async Task<IEnumerable<IConnectorContainer>> GetContainers(ExecutionContext executionContext, Guid connectorProviderDefinitionId)
```
[the remark this answers is not in the document]
Comment (AndreasNicolaisen) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:465]: Maybe we should, I am not sure what this the result of this function is used for. In platform, its only usage is in [`ConnectorController`](https://github.com/CluedIn-io/CluedIn/blob/develop/Code/Server.WebApi/Controllers/Consume/Connectors/ConnectorController.cs#L41), were it simply returns the result. I can't find any place in the UI were it's used, so can't tell for sure what is expected.

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs [hunk 5]
```diff
@@ -22,546 +16,463 @@
...
+            await ExecuteWithRetryAsync(async () =>
+            {
+                var config = await AuthenticationDetailsHelper.GetAuthenticationDetails(executionContext, streamModel.ConnectorProviderDefinitionId.Value);
+                await using var connectionAndTransaction = await _client.BeginTransaction(config.Authentication);
+                var transaction = connectionAndTransaction.Transaction;
 
-        public Task<string> GetCorrelationId()
-        {
-            return Task.FromResult(Guid.NewGuid().ToString());
+                await transaction.CommitAsync();
```
[the remark this answers is not in the document]
Comment (zvirja) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:45]: Should we than run some dummy query? Like `SELECT 1 FROM <table>` or similar?

[the remark this answers is not in the document]
Comment (AndreasNicolaisen) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:45]: Good idea, added dummy select query from main table.

File: src/Connector.SqlServer/Connector/SqlServerConnector.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (AndreasNicolaisen) [reply src/Connector.SqlServer/Connector/SqlServerConnector.cs:471]: Added filtering down to only yield main table. The way I'm doing this, is to find tables, that have the columns, that a main table as a minimum would have. If we change what those are, we might need to rework this.

**Review by Hyldahl — APPROVED** [review]
Looks good 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

**Review by zvirja — APPROVED** [review]

**Review by zvirja — APPROVED** [review]

**Pull request merged**